src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -97,31 +113,65 @@ new fs hbio mergeCaching alloc mergeLevel mergeMappend targetPaths runs = do
       let numEntries = coerce (sum @V.Vector @Int) (fmap Run.runNumEntries runs)
       mergeBuilder <- Builder.new fs targetPaths numEntries alloc
-      mergeLastStepResult <- newMutVar $! MergeInProgress
+      mergeState <- newMutVar $! Merging
+      mergeRefCounter <-
+        RC.mkRefCounter1 (Just $! finaliser fs hbio mergeState mergeBuilder mergeReaders)
       return Merge {
           mergeHasFS = fs
         , mergeHasBlockIO = hbio
         , ..
         }
 
-{-# SPECIALISE close :: Merge IO (FS.Handle h) -> IO () #-}
--- | This function should be called when discarding a 'Merge' before it
--- was done (i.e. returned 'MergeComplete'). This removes the incomplete files
--- created for the new run so far and avoids leaking file handles.
+{-# SPECIALISE addReference :: Merge IO h -> IO () #-}
+addReference :: (HasCallStack, PrimMonad m) => Merge m h -> m ()
+addReference Merge{..} = RC.addReference mergeRefCounter
+
+{-# SPECIALISE removeReference :: Merge IO h -> IO () #-}
+removeReference :: (HasCallStack, PrimMonad m, MonadMask m) => Merge m h -> m ()
+removeReference Merge{..} = RC.removeReference mergeRefCounter
+
+{-# SPECIALISE finaliser :: HasFS IO h -> HasBlockIO IO h -> MutVar RealWorld MergeState -> RunBuilder RealWorld (FS.Handle h) -> Readers IO (FS.Handle h) -> IO () #-}
+-- | Closes the underlying builder and readers.
 --
--- Once it has been called, do not use the 'Merge' any more!
-close :: (MonadFix m, MonadSTM m, MonadST m) => Merge m h -> m ()
-close Merge {..} = do
-    Builder.close mergeHasFS mergeBuilder
-    Readers.close mergeHasFS mergeHasBlockIO mergeReaders
+-- This function is idempotent. Technically, this is not necessary because the
+-- finaliser is going to run only once, but it is a nice property for
+-- @close@-like functions to be idempotent.
+finaliser ::
+     (MonadFix m, MonadSTM m, MonadST m)
+  => HasFS m h
+  -> HasBlockIO m h
+  -> MutVar (PrimState m) MergeState
+  -> RunBuilder (PrimState m) (FS.Handle h)
+  -> Readers m (FS.Handle h)
+  -> m ()
+finaliser hfs hbio var b rs = do
+    st <- readMutVar var
+    let shouldClose = case st of
+          Merging     -> True
+          MergingDone -> True
+          Completed   -> False
+          Closed      -> False
+    when shouldClose $ do
+        Builder.close hfs b
+        Readers.close hfs hbio rs
+        writeMutVar var $! Closed
 
 {-# SPECIALISE complete ::
      Merge IO h
   -> IO (Run IO (FS.Handle h)) #-}
 -- | Complete a 'Merge', returning a new 'Run' as the result of merging the
--- input runs. This function will /not/ do any merging work if there is any
--- remaining. That is, if not enough 'steps' were performed to exhaust the input
--- 'Readers', this function will throw an error.
+-- input runs.
+--
+-- The resulting run has the same reference count as the input 'Merge'. The
+-- 'Merge' does not have to be closed afterwards, since it is closed implicitly
+-- by 'complete'.
```

Comment:
I wonder if `close` should not explicit, even after `complete`. It is more intuitive for the user/reader of this code if fewer things are implicit.  A similar comment applies to `RunBuilder`, which also performs an implicit `close` on `unsafeFinalise`

Reply:
(not in scope for this PR, though)

Reply:
I've made a note for myself to come back to this at some point. Thanks for your input @recursion-ninja